### PR TITLE
Implement Start Timer Event. Resolves #25

### LINKED ISSUE #25: Implement Start Timer Event

### PR DESCRIPTION
**Pull request by caleeli**
This pull request implements TimerEventDefinitions in Start Events.
image
It implements:

- Start a process by a ISO8601 specific date time expression
- Start a process cyclically by a ISO8601 specific cyclic expression
- Start a process by a calculable expression 
- Start a process cyclically by a calculable expression
image
image

### CODE REVIEW

File: src/ProcessMaker/Nayra/Bpmn/IntermediateCatchEventTrait.php [hunk 1]
```diff
@@ -120,5 +121,10 @@ public function execute(EventDefinitionInterface $message, ExecutionInstanceInte
         $this->triggerPlace->addNewToken($instance);
         return $this;
     }
+
```
Comment (caleeli) [inline src/ProcessMaker/Nayra/Bpmn/IntermediateCatchEventTrait.php]: Add a register method for the catch event elements like intermediate catch events.

File: src/ProcessMaker/Nayra/Bpmn/StartEventTrait.php [hunk 2]
```diff
@@ -106,4 +107,18 @@ public function execute(EventDefinitionInterface $event, ExecutionInstanceInterf
         }
         return $this;
     }
+
+    /**
```
Comment (caleeli) [inline src/ProcessMaker/Nayra/Bpmn/StartEventTrait.php]: Add a register method for the catch event elements like start events.

File: src/ProcessMaker/Nayra/Bpmn/TimerEventDefinition.php [hunk 3]
```diff
@@ -0,0 +1,162 @@
+<?php
```
Comment (caleeli) [inline src/ProcessMaker/Nayra/Bpmn/TimerEventDefinition.php]: Message event definition class for the timers.

File: src/ProcessMaker/Nayra/Bpmn/TimerEventDefinition.php [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    /**
+     * Schedule as timeDuration.
+     *
+     * @param EngineInterface $engine
+     * @param FlowElementInterface $element
+     * @param TokenInterface $token
+     *
+     * @codeCoverageIgnore Until intermediate timer event implementation
```
Comment (caleeli) [inline src/ProcessMaker/Nayra/Bpmn/TimerEventDefinition.php]: Note that a timer event of type duration, is not posible in a start event.

File: src/ProcessMaker/Nayra/Contracts/Bpmn/TimerEventDefinitionInterface.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<?php
```
Comment (caleeli) [inline src/ProcessMaker/Nayra/Contracts/Bpmn/TimerEventDefinitionInterface.php]: This interface defines the timer event definition properties and events

File: src/ProcessMaker/Nayra/Contracts/Engine/JobManagerInterface.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
+<?php
```
Comment (caleeli) [inline src/ProcessMaker/Nayra/Contracts/Engine/JobManagerInterface.php]: This interface defines the different types of schedules that a timer event could use.

File: src/ProcessMaker/Nayra/Engine/EngineTrait.php [hunk 7]
```diff
@@ -2,6 +2,7 @@
```
Comment (caleeli) [inline src/ProcessMaker/Nayra/Engine/EngineTrait.php]: This change add support to load processes without executing them into the engine. The execution will be delayed and controlled by the start timer event.

File: tests/Feature/Engine/EngineTestCase.php [hunk 8]
```diff
@@ -4,9 +4,12 @@
```
Comment (caleeli) [inline tests/Feature/Engine/EngineTestCase.php]: Added a mock of jobs manager

File: tests/Feature/Engine/StartTimerEventTest.php [hunk 9]
```diff
@@ -0,0 +1,190 @@
+<?php
```
Comment (caleeli) [inline tests/Feature/Engine/StartTimerEventTest.php]: This file implements the tests for the start timer event

File: tests/ProcessMaker/Models/FormalExpression.php [hunk 10]
```diff
@@ -2,49 +2,112 @@
```
Comment (caleeli) [inline tests/ProcessMaker/Models/FormalExpression.php]: This change add support for timer expressions.

File: src/ProcessMaker/Nayra/Bpmn/IntermediateCatchEventTrait.php [hunk 11]
```diff
@@ -120,5 +121,10 @@ public function execute(EventDefinitionInterface $message, ExecutionInstanceInte
...
+    public function registerCatchEvents(EngineInterface $engine)
```
Comment (danloa) [inline src/ProcessMaker/Nayra/Bpmn/IntermediateCatchEventTrait.php]: add documentation for this function

File: src/ProcessMaker/Nayra/Contracts/Bpmn/CatchEventInterface.php [hunk 12]
```diff
@@ -23,4 +25,6 @@ public function getEventDefinitions();
      * @return \ProcessMaker\Nayra\Engine\ExecutionInstance[]
      */
     public function getTargetInstances(EventDefinitionInterface $message, TokenInterface $token);
+
+    public function registerCatchEvents(EngineInterface $engine);
```
Comment (danloa) [inline src/ProcessMaker/Nayra/Contracts/Bpmn/CatchEventInterface.php]: Add documentation to  this method signature

File: src/ProcessMaker/Nayra/Contracts/Engine/JobManagerInterface.php (hunk 6 above)
Comment (danloa) [reply src/ProcessMaker/Nayra/Contracts/Engine/JobManagerInterface.php]: Wouldn't it be better to add this git comment as part of the interface comment?

File: tests/Feature/Engine/StartTimerEventTest.php [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+    {
+        //Load a BpmnFile Repository
+        $bpmnRepository = new BpmnFileRepository();
+        $bpmnRepository->setEngine($this->engine);
+        $bpmnRepository->load(__DIR__ . '/files/Timer_StartEvent_TimeCycleExpression.bpmn');
+
+        //Create a default environment data
+        $environmentData = $bpmnRepository->getDataStoreRepository()->createDataStoreInstance();
+        $this->engine->setDataStore($environmentData);
+        $everyMinute = '1M';
```
Comment (danloa) [inline tests/Feature/Engine/StartTimerEventTest.php]: why use a variable and not just add the string to the expression below?

File: src/ProcessMaker/Nayra/Contracts/Engine/JobManagerInterface.php (hunk 6 above)
Comment (caleeli) [reply src/ProcessMaker/Nayra/Contracts/Engine/JobManagerInterface.php]: Yes

File: src/ProcessMaker/Nayra/Contracts/Bpmn/CatchEventInterface.php (hunk 12 above)
Comment (caleeli) [reply src/ProcessMaker/Nayra/Contracts/Bpmn/CatchEventInterface.php]: Completed

File: src/ProcessMaker/Nayra/Bpmn/IntermediateCatchEventTrait.php (hunk 1 above)
Comment (caleeli) [reply src/ProcessMaker/Nayra/Bpmn/IntermediateCatchEventTrait.php]: Completed

File: tests/Feature/Engine/StartTimerEventTest.php (hunk 13 above)
Comment (caleeli) [reply tests/Feature/Engine/StartTimerEventTest.php]: To distinguish the periodic value. I put a comment to clarify that

**Comment by caleeli** [comment]
@danloa As suggested in the CR, it was completed the code comments.

**Pull request merged**